### PR TITLE
builder: Add `ServiceBuilder::service_fn`

### PR DESCRIPTION
**Pull request by davidpdrsn**
A small convenience for doing `.service(service_fn(handler_function))`.

### CODE REVIEW

**Review by hawkw — APPROVED** [review]

File: tower/src/builder/mod.rs [hunk 1]
```diff
@@ -459,6 +459,22 @@ impl<L> ServiceBuilder<L> {
     {
         self.layer.layer(service)
     }
+
+    /// Wrap the async function `F` with the middleware provided by this [`ServiceBuilder`]'s
+    /// [`Layer`]'s, returning a new [`Service`].
+    ///
+    /// This is a convenience for doing `.service(service_fn(handler_function))`.
+    ///
```
Comment (hawkw) [inline tower/src/builder/mod.rs:467]: nit: might be nice to have an example?

File: tower/src/builder/mod.rs [hunk 2]
```diff
@@ -459,6 +459,22 @@ impl<L> ServiceBuilder<L> {
     {
         self.layer.layer(service)
     }
+
+    /// Wrap the async function `F` with the middleware provided by this [`ServiceBuilder`]'s
+    /// [`Layer`]'s, returning a new [`Service`].
+    ///
+    /// This is a convenience for doing `.service(service_fn(handler_function))`.
```
Comment (hawkw) [inline tower/src/builder/mod.rs:466]: nit: might be nice if this linked to `service_fn` someplace?

File: tower/src/builder/mod.rs (hunk 1 above)
Comment (davidpdrsn) [reply tower/src/builder/mod.rs:491]: Sure thing!

File: tower/src/builder/mod.rs (hunk 2 above)
Comment (davidpdrsn) [reply tower/src/builder/mod.rs]: Good idea. I made `.service(service_fn(handler_function))` link to it. Doesn't think that totally makes sense but not sure where else to put it 🤷

File: tower/src/builder/mod.rs [hunk 3]
```diff
@@ -459,6 +459,47 @@ impl<L> ServiceBuilder<L> {
     {
         self.layer.layer(service)
     }
+
+    /// Wrap the async function `F` with the middleware provided by this [`ServiceBuilder`]'s
+    /// [`Layer`]'s, returning a new [`Service`].
+    ///
+    /// This is a convenience for doing [`.service(service_fn(handler_function))`][`service_fn`].
```
Comment (hawkw) [inline tower/src/builder/mod.rs:466]: nit, take it or leave it: maybe something like this is slightly nicer:
```suggestion
    /// This is a convenience method which is equivalent to calling
    /// [`ServiceBuilder::service`] with a [`service_fn`], like this:
    /// ```rust
    /// # use tower::{ServiceBuilder, ServiceFn};
    /// # async fn handler_fn(_: ()) -> Result<(), ()> { Ok(()) }
    /// # let _ = {
    /// ServiceBuilder::new()
    ///     // ...
    ///     .service(service_fn(handler_fn))
    /// # };
```

File: tower/src/builder/mod.rs [hunk 4]
```diff
@@ -459,6 +459,47 @@ impl<L> ServiceBuilder<L> {
...
+    ///
+    /// let response = svc.oneshot("foo").await?;
+    ///
+    /// assert_eq!(response, "foo");
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// [`Layer`]: crate::Layer
+    /// [`Service`]: crate::Service
```
Comment (hawkw) [inline tower/src/builder/mod.rs:493]: do these links actually work with the reexports, or do we need to link to the original crates? could be good to test that out in a docs build.

**Review by hawkw — APPROVED** [review]
I commented on some potential docs improvements, but other than that, this looks great!

File: tower/src/builder/mod.rs (hunk 4 above)
Comment (hawkw) [reply tower/src/builder/mod.rs:493]: again, turns out we're already checking intra-docs links on CI, and i don't need to worry about this, lol.

File: tower/src/builder/mod.rs (hunk 3 above)
Comment (davidpdrsn) [reply tower/src/builder/mod.rs:466]: Much nicer!

**Pull request merged**